Added port to Python3 of htmresearch location framework to htm.core

fcr opened the pull request:
I'm in the middle of porting selected Python parts of htmresearch to htm.core and Python 3.
I needed to check that the Connections were working so added some tests.
I may expand these in the future if the need arises.

breznak commented:
Thanks, there's a previous work in this matter, #349 , how does that compare? 

A word of caution, Connections are considered our internal, backend structure, so are freely subject to change, even API breakage. 
On the other hand I assume this work is for htm.reseach, so you would be using Connections directly (just keep in mind that same as research is not stable, nor is Connections). 

But if we already have a py bindings for connections, we shoould also have tests.

fcr commented:
I have ported the location framework from htmresearch so that it is based on htm.core rather than nupic.core and runs under Python3.
The location framework is the code that underlies Jeff Hawkins Thousand Brains model of the Neocortex and hence is probably the most important framework in htmresearch.
The port includes unit tests and a couple of example applications using the location framework.
This pull request also includes a few additions the the htm.core Python bindings that are needed to support the location framework port.
Please note that there is a subclass of the Connections class which provides some methods that are missing from the htm.core Connections.cpp.  These methods would be good candidates for inclusion in the Connections.cpp.

breznak commented:
> I have ported the location framework from htmresearch so that it is based on htm.core rather than nupic.core and runs under Python3.
The location framework is the code that underlies Jeff Hawkins Thousand Brains model of the Neocortex and hence is probably the most important framework in htmresearch.

@fcr This is an awesome feat! :100: 

Before I start looking into this, and I'd like both @dkeeney , @ctrl-z-9000-times to make their reviews too, as this is an important update! ... this is 17k LOC, do you think you could split this into a series of more digestable PRs? (if not, we'll go through, but modularization would be welcomed)

Maybe?: 

> The port includes unit tests and a couple of example applications using the location framework.
This pull request also includes a few additions the the htm.core Python bindings that are needed to support the location framework port. 

PR-1

> Additional goodies are the RawSensor and RawValues regions, the GridCellLocationRegion and the ColumnPoolerRegion.

Raw, GridCells, CP as separate PRs 2,3,4?

breznak commented on bindings/py/cpp_src/bindings/engine/py_Engine.cpp:
```diff
@@ -388,7 +390,8 @@ namespace htm_ext
             .def("getMinEnabledPhase", &htm::Network::getMinPhase)
             .def("getMaxEnabledPhase", &htm::Network::getMaxPhase)
-			.def("run",                &htm::Network::run);
+            .def("setPhases",          &htm::Network::setPhases)
```
what are Phases used for?

fcr replied:
Phases are uses to determine the order that regions compute their input.

breznak commented on py/htm/algorithms/apical_tiebreak_temporal_memory.py:
```diff
@@ -0,0 +1,1126 @@
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""An implementation of TemporalMemory"""
```
- [ ] please add a note to files that are ported from htm.research that they come from there. 
- [ ] mention that the API may break (from research), you may add a note how lively/stable the code is now
- [ ] use a separate "namespace" for these. py/htm/{experimental, research?}/algorithms

fcr replied:
- [ ] Okay I will add the note. Should have done that from the beginning.
- [ ] I don't want to use the namespace "htmresearch" since I consider these as being usable and Numenta is no longer supporting them. 
- [ ] I cannot see these API's breaking for the same reason.
- [ ] I suggest the namespace "htm_advanced" since this is what they really are. Also this is the namespace that I used when developed the port so it will be quicker to make the change back.

breznak replied:
> as being usable and Numenta is no longer supporting them.

interesting point. Numenta is abandoning their whole htmresearch repo? Due to the py3 switch their trying to pull? 

>  I cannot see these API's breaking for the same reason.

ok, good idea would be to make a `htm_advanced/README.md` file and mention all these things:
- that the files are ported from htmresearch repo
- that (and why?) Numenta is no longer developing them
  - comes with API being stable
  - note, the files directly use Connections. We consider connections our internal code, so the APIs may break, but not likely. 
  - htm_advanced is fine, esp if it'd be easy for you to revert to that. Otherwise I'd vote only for "advanced", or some other names.

rhyolight replied:
> Numenta is abandoning their whole htmresearch repo? Due to the py3 switch their trying to pull?

I wouldn't say we are abandoning it. We'll port the code we need again to py3 as we need it. These implementations in Py2 are all we have, but we probably won't be running python2 runtimes anymore. So they are here until we need to port them (as @fcr just did -- thanks, btw!)

breznak replied:
Thanks for the clarification, Matt!

fcr commented:
[@dkeeney](https://github.com/dkeeney)> As always, I have to defer to others regarding the usefulness of the algorithms. I was hoping to find the NetworkAPI Region modules for the algorithms. I would like my NetworkAPI users to have access to these new algorithms. Do they exist? Are they coming in a later PR or am I looking in the wrong place. Since you are using a personal fork I cannot see your entire branch.

If I understand your question, look in the htm.regions package. There are quite a few new regions.
> 
> Otherwise, I have no problem with the changes that I can see.

ctrl-z-9000-times commented:
Do you need to include the `*.training.log` files?  They appear to be just lists of numbers.

breznak commented:
> Do you need to include the *.training.log files? They appear to be just lists of numbers.

nice, that would significantly bring down the num of changed files

dkeeney commented:
> If I understand your question, look in the htm.regions package. There are quite a few new regions.

Ah, I figured it out.  I have to click on your branch name at the top in order to see everything.
And yes.  There are the Regions.  Wonderful.

fcr commented:
> Do you need to include the `*.training.log` files? They appear to be just lists of numbers.

Yes. They are the object definition data for the example.

fcr commented:
BTW are you still sure that you want me to separate out the additions. Now that I have done it (but will not have time to release it until some time tomorrow - I am in the Israel time zone), it just doesn't look so nice.
algorithms in htm, algorithms in htm_advanced, examples under htm, examples under htm_advanced and two levels of tests.

breznak commented:
> to separate out the additions. [...] it just doesn't look so nice:
algorithms in htm, algorithms in htm_advanced, examples under htm, examples under htm_advanced etc. and two levels of tests.

depends if it's more readable in logical steps..?

fcr commented:
Gentlemen,
Made the requested changes:

- [ ] All the location framework packages and the examples are in the htm_advance package. All the tests are in tests/htm_advanced.
- [ ] All unit tests, cpp and Python pass except for HelloSPTPTest.performance. 
- [ ] There is a readme in htm_advanced

A reminder that the pull request contains changes to some cpps and bindings that the framework required.

And on a closing note, so that it does not get lost in the thread, once again please note that there is a Python subclass of the Connections class which provides some methods that are missing from the htm.core Connections.cpp. These methods would be good candidates for inclusion in the Connections.cpp.

Okay, its all yours. Enjoy.

dkeeney commented:
@fcr This is a major achievement.  Good work.

Oh, all three CI's are failing in column_pooler_test.py on lines 23 and 26.
```
    from prettytable import PrettyTable
E   ModuleNotFoundError: No module named 'prettytable'
```
Is this a package we need?
All three CI's must turn green before it will let us merge into master.

fcr commented:
> ```
>     from prettytable import PrettyTable
> E   ModuleNotFoundError: No module named 'prettytable'
> ```
> 
> Is this a package we need?
> All three CI's must turn green before it will let us merge into master.

Yes. I didn't realize that PrettyTable was not a standard package. I've had it for ages.
Will it be a problem to make PrettyTable a requirement?
If it is a problem I can safely remove it at the cost of some useful functionality.

breznak commented:
> Will it be a problem to make PrettyTable a requirement?

if it can be installed by `pip` for all platforms, no problem at all. 

> And on a closing note, so that it does not get lost in the thread, once again please note that there is a Python subclass of the Connections class which provides some methods that are missing from the htm.core Connections.cpp. These methods would be good candidates for inclusion in the Connections.cpp.

Sounds reasonable, I haven't seen the mentioned methods yet, if they are useful in proader context. But likely this could be a follow-up issue. 

> unit tests, cpp and Python pass except for HelloSPTPTest.performance.

this would be strange, you haven't touched the core c++ which could affect performance (?) 

Thanks for the new changes!

breznak commented on py/htm_advanced/examples/thing_classification/data/002_master_chef_can.training.log:
```diff
@@ -0,0 +1,26 @@
+[(68.1235, -9.7628, 0.7851)/10] => [[5,26,35,68,88,97,328,396,421,466,545,587,608,635,693,732,774,913,964],[91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131]]
```
what are these files, their format? Looks like some sensory-motor data for objects? Please add a README in the `/data/`

fcr replied:
Same comment as for the howtos. I want to tackle the metric classes first.

dkeeney commented:
> unit tests, cpp and Python pass except for HelloSPTPTest.performance. 

The performance tests fail on my machines sometimes too.  Both Windows and Ubuntu.

dkeeney commented:
> Will it be a problem to make PrettyTable a requirement?

If you decide to make this a requirement, you will need to add it to `bindings/py/packaging/requirements.txt`  so that everyone gets it.

fcr commented on py/htm_advanced/algorithms/monitor_mixin/monitor_mixin_base.py:
```diff
@@ -0,0 +1,223 @@
+1. Create a variable for the data you want to capture in your specific monitor's
+`mmClearHistory` method. For example,
+
+        self._mmTraces["predictedCells"] = IndicesTrace(self, "predicted cells")
```
For my own usage, I wanted to get the framework functionally working as close as possible to the original htmresearch implementation so I was not particularly critical of the measurement code as long as everything worked.
I didn't notice the Metrics capability in the bindings. Now that you brought my attention to it, I'll have a look and see what is possible.

ctrl-z-9000-times replied:
I don't think this can be replaced with the existing SDR Metrics class.  This code appears to keep a full history of the data, so that it can do a more thorough analysis (plot & histogram).  The existing metrics class does a limited analysis on the data when it is received, but it does not save the data.  This difference is critical: long running applications can not save every bit of data forever, and lab-experiments need to save every bit of data.

On the whole, I think this new metrics class looks useful and I hear that it works.  It's also baked into the research code base, so refactoring it is non-trivial.

fcr replied:
> On the whole, I think this new metrics class looks useful and I hear that it works. It's also baked into the research code base, so refactoring it is non-trivial.

It definitely works. The example applications use it heavily.

breznak replied:
Sounds fair, let's keep both metrics as is. 

> I wanted to get the framework functionally working as close as possible to the original htmresearch implementation so I was not particularly critical of 

actually, this is what I'm trying to figure for this PR, be happy that it works and looks as closely as the original repo, or look more at integration with the current code..

fcr replied:
It is literally a straight Python2 to Python3 port but based on htm.core rather than nupic.core and with no embellishments.
The unit tests are the original unit tests from htmresearch (ported of course)

fcr commented on py/htm_advanced/README.md:
```diff
@@ -0,0 +1,24 @@
+
+Location Framework
+==============
+The location framework is the code that underlies Jeff Hawkins Thousand Brains model of the Neocortex and hence is probably the most important framework in htmresearch.  
```
Okay. I've go this far I might as well move it to htm.advanced. My git is going to have hissy fit. :-)
I'll scratch all your nits.

fcr commented on py/htm_advanced/algorithms/connections.py:
```diff
@@ -0,0 +1,155 @@
+from htm.bindings.algorithms import Connections as CPPConnections
+import numpy as np
+
+class Connections(CPPConnections):
```
Wouldn't it be nicer if you made it unnecessary?

breznak replied:
It definitely would, but personally I know I won't have time to do that in near future. And it likely should be addressed as a separate issue, followup anyway, this PR is already a huge feat! :+1:

breznak replied:
Now tracked in #668

fcr commented:
Refactored the namespace yet again. The code is now in htm.advanced as per nit that was picked. :-)
Added PrettyTable to requirements.txt
And even added a ReadMe to the experiments that was lacking one. ;-)

fcr commented:
I just realized that I could use the `Region.executeCommand()` API to access methods in Region Python subclasses. I hadn't noticed that executeCommand can pass arguments. 
I can use it to access the addToQueue for RawSensor and RawValue regions, rather than the current inelegant implementation via setParamterArray!
I tested it and it works. :-)
I'll make the change plus some other small improvements (as separate pull requests) once (if) you accept the current pull request.

breznak commented:
> I can use it to access the addToQueue for RawSensor and RawValue regions, rather than the current inelegant implementation via setParamterArray!
I tested it and it works. :-)
I'll make the change plus some other small improvements (as separate pull requests)

@dkeeney can you please have a look at the Region* classes added? 
@fcr  Hey, this is great find! if this new thing works, please add it to the current PR. 

I'm definitely going to accept the PR, maybe today or during the weekend (I'd like to give some time to the Davids to voice their ideas/requests). I just want to read through the new classes, so I understand them. I won't be expecting any wild changes/requests to them, since this is a port from py2. 

- [ ] one thing I want to make sure is that (most) of the code is run in unittests as a part of the CI (monitors, apicalTM, examples,.. ) For the code that is run, we'll catch API breaks.

fcr commented:
I will not have time until Sunday to propagate the change to all places using addToQueue. So you can either wait until then to accept the current pull request or I just make a new pull request for this change later.

dkeeney commented:
@fcr the executeCommand facility is not exactly a direct call into your Region Impl but it is probably the best you can do. There are some limitations on the types of arguments you can use and the Region impl needs to have code to handle the executeCommand request.  

For C++ apps there is also a callback facility that lets you register a function to be called at the end of each run cycle.  That is where you can load up your buffers for the next cycle. What might be useful is some way that a python app could register something too but I have not looked into that yet.

fcr commented:
>[@dkeeney](https://github.com/dkeeney)Region impl needs to have code to handle the executeCommand request.
That's great. It gives full control to the Region implementation.

fcr commented:
I have implemented and released the change from setParameterArray to executeCommand. 
The resultant code is much nicer IMHO. 
Note that region.__init__.py has some helper methods for the executeCommand to make thinngs neater.
I also took the opportunity to fix some minor problems and add some new unit tests.

ctrl-z-9000-times commented:
Hi fcr,

I'm trying to run the examples but I can't seem to get them to work.  What I'd really like to see are the figures from the columns papers.  Can you demonstrate that this can reproduce any of the images?

Thanks.

fcr commented:
[@ctrl-z-9000-times](https://github.com/ctrl-z-9000-times) Did you run them with the committed in configurations? Anything less will not accumulate the statistics needed to generate the figure correctly if at all.
But when you do that, they take ages to run. 
I did run them earlier and got the figures but I deleted the results folders before committing my push request to keep the release clean.
After getting your query, I started running the union path integration example directly in my copy of htm.core. It is still merrily chugging away and will probably take most of the night.
In any case my main motivation for including the examples was to show example usage of the framework and as a sort of final sanity check - not so much for people to reproduce the paper results since this takes so long.
I'll let you know what I get when I have run both examples again.

ctrl-z-9000-times commented:
> @ctrl-z-9000-times Did you run them with the committed in configurations? 

I did, but I did not wait very long.  If the results are the same as the Numenta papers results, then I guess there is nothing to see :)  If you know the results are the same then I'm on board with this PR.  Thank you for re-running this but you do not need to re-run it if it works.

+1

fcr commented:
union path integration was finished when I woke up this morning. About 1.3Gb of plots and graphs. I can send them if you want  - they are only 195Mb compressed. Just don't accuse me of sending zip bombs. :-). 
Since you are convinced, I will not rerun the thing calculation. It takes even longer.

breznak commented:
> After getting your query, I started running the union path integration example directly in my copy of htm.core. It is still merrily chugging away and will probably take most of the night.

thank you both for throuroughly testing the code, and (re)running the examples. Good to know everything works (tm). Out of curiosity, are the results with htm.core somewhat significanltly different from those based on nupic.core? (I wouldn't think so, but we do some computations slightly differently)

breznak commented on py/htm/advanced/regions/RawSensor.py:
```diff
@@ -19,8 +19,8 @@
-from htm.bindings.engine_internal import Array
-import numpy as np
+
+from .import extractList
```
nit, probably ok, but next time please use the absolute paths for imports. The code is then easier to read and move around.

fcr replied:
The reference emphasizes the close relationship of the method with the region implementations.  Its there to support them. Besides this is good Pythonic 3 usage. :-)

breznak commented on py/htm/advanced/regions/__init__.py:
```diff
@@ -0,0 +1,36 @@
+
+from collections.abc import Iterable
+
+def executeCommand(command_name, region_instance, *args):
```
is this the "enablement" you've discussed with @dkeeney ? This code looks quite generic, would it be possible to add it to the "main" engine (or relevant place)?

fcr replied:
Yes. It is generic but its home should be in Python because it provides a generic entry for expanding the API of PyRegion implementations.

dkeeney replied:
> would it be possible to add it to the "main" engine 

This is in the main engine in the sense that executeCommand( ) is part of the API and is available for use by C++ region implementations.  What he did was expose it to python.  That was an over-site on my part.  It should have been there all along.  Thank you @fcr.

breznak replied:
> executeCommand( ) is part of the API and is available for use by C++ region implementations. What he did was expose it to python

yes, yet here it's still under `htm.advanced.regions` what I'm trying to figure is, if we can add it to `PyRegionImpl` (in the bindings for a generic Region class, or some other suitable place)?

dkeeney replied:
@breznak  Oh, I total missed the point you were making.  I had to go look at this again.

Yes, I agree, we should make those changes available to all python users, not just those using the advanced section.  The C++ signature for the call is:
```
std::string executeCommand(const  std::string& command, vector<std::string>)
```
If I am following this correctly, @fcr is making a python function that accepts a normal argument list and creates a vector of strings so it will match the C++ signature.

What we probably should do is implement that in the bindings. `bindings/py/cpp_src/bindings/engine/py_Engine.cpp`   in the `Region` section where it exposes the executeCommand.   We probably want to keep the existing calling signature and add a new overload that accepts *args rather than the vector.

@fcr, I can help with that if you do not know how to do it.

fcr replied:
My C++ skills are fairly weak so yes, it would be better if you were to do this.

dkeeney replied:
My Python skills are fairly weak to together we could make a good team :-)
Ok, let me make a new PR.  Then when that is merged we can try your PR to make sure this works with your PR and then your PR would be ready to merge (if all else is ok).

breznak replied:
Tag me David, and I'll merge the PR asap

fcr commented:
Pushed another another small bug that I found and blocked my progress.

fcr commented:
Oops. Forgot to run the C++ unit test before pushing. :-(

breznak commented on src/htm/regions/SPRegion.cpp:
```diff
@@ -349,7 +349,7 @@ Spec *SPRegion::createSpec() {
                     "",                               // constraints
                     "0.1",                            // defaultValue
-                    ParameterSpec::ReadOnlyAccess)); // access
+                    ParameterSpec::ReadWriteAccess)); // access
```
Please don't. Sorry it isn't really clear from the current code, it is a WIP in #561 , `synPermConnected` is essentially const, since it's used in Connections during initialization (and cannot, by its logic, be changed after that). A code that would change that variable would break all previous learning. 

You mentioned this was "another obstacle" for you, how is that?

fcr replied:
I need to be able to set this in creating my SPRegion or is there another way of doing this?

fcr replied:
I guess I'll meanwhile rollback this change and get your help to deal with this issue.

breznak replied:
> I need to be able to set this in creating my SPRegion or is there another way of doing this?

You can create a SPRegion, just don't set the RO variables. 
If you wanted to change the synPermConnected after initialization of the region, don't. The previous learning will be lost, and the code shouldn't allow it.

fcr replied:
No. I want to set the synPermConnected at SPRegion creation. This is normally done by means of the parameters dict. However, synPermConnected has ReadOnlyAccess and hence that entry in the parameter dict is rejected.

fcr replied:
Maybe the bug is that the initialization parameter being rejected.

breznak replied:
> I want to set the synPermConnected at SPRegion creation. This is normally done by means of the parameters dict. However, synPermConnected has ReadOnlyAccess and hence that entry in the parameter dict is rejected.

FYI @dkeeney , I think this is a limitation of the NetworkAPI (?). We want to be able to set some variables during initializer-time (equivalent of `SP(..., synPermConnected=0.222)` but the var. should be read-only after that during runtime (so `sp.setSynPermConnected(xxx)` is not allowed). 

@fcr you could workaround the inability to change synPermConnected by appropriately changing `synPermActiveInc/Dec`, it has the same effect.

fcr replied:
>FYI @dkeeney , I think this is a limitation of the NetworkAPI (?). We want to be able to set some variables during initializer-time (equivalent of SP(..., synPermConnected=0.222) but the var. should be read-only after that during runtime (so sp.setSynPermConnected(xxx) is not allowed).

Exactly. 
This is a serous bug IMHO.
Because of this limitation, I had to set the access for all the parameters in the advanced/regions to ReadWrite in order to allow initialization of them, even though they should not be altered afterwards.

dkeeney replied:
The Spec should reflect what the code implements.  So just changing the spec would allow the setParameterXXX to be called but it would probably not be handled by SPRegion unless you change the setParameterXXX( ) function as well to handle it.  But if the algorithm cannot accept it as @breznak mentioned then you would probably break something.

By the way.  A parameter passed in during creation should have an access of CreateAccess rather than ReadOnlyAccess.  CreateAccess means can be used for initialization but read-only after that.  ReadOnlyAccess implies that it is something created by the algorithm and is readable.

fcr replied:
>@fcr you could workaround the inability to change synPermConnected by appropriately changing synPermActiveInc/Dec, it has the same effect.

Same effect but wrong semantics.

fcr replied:
> By the way. A parameter passed in during creation should have an access of CreateAccess rather than ReadOnlyAccess. CreateAccess means can be used for initialization but read-only after that. ReadOnlyAccess implies that it is something created by the algorithm and is readable.

Great! 
Can you please change synPermConnected to CreateAccess?
I'm not touching the C++ stuff any more. Don't want to break anything. :-)

fcr replied:
Once this PR is finally accepted, I'll go through and start applying the CreateAccess to the relevant region specs and make a new (smaller) PR.

fcr replied:
BTW never came across CreatAccess before.

breznak replied:
tracked in #669

fcr commented:
I have restored synPermConnected to ReadOnlyAccess, but I have left the test for this py/tests/regions/sp_region_test.SPRegionTest.testSPRegionParametersAreWritable() but it is skipped for now.
If you unskip it you will understand what I mean.

fcr commented:
See my last comment in #669

breznak commented:
@fcr Can you please merge master with #670 , which includes David's fix for the issue with `executeCommand`? And it should allow you to get rid of the custom `htm.advanced.regions.__init__` `executeCommand`

breznak commented on py/htm/advanced/support/register_regions.py:
```diff
@@ -0,0 +1,55 @@
+# ----------------------------------------------------------------------
+"""
+Method to register regions in the nupic.research repository. Regions
+specified here must be registered before NuPIC's Network API can locate them.
```
can this be moved to `htm.advanced.regions.__init__` ?

breznak commented on py/htm/advanced/regions/RawSensor.py:
```diff
@@ -0,0 +1,169 @@
+    in a FIFO and each call to compute pops the top element.
+
+    Each data record consists of the non-zero indices of the sparse vector,
+    a 0/1 reset flag, and an integer sequence ID.
```
@dkeeney do we need this region? "sparse vector" is exactly our SDR, and I think the main InputRegion was adapted to accept SDRs? 
Q is if it supports the additional bells&whistles (0/1 reset flag, and sequence ID)?

dkeeney replied:
> @dkeeney do we need this region? "sparse vector" is exactly our SDR,

I think that this is not so much a facility to create a sparse vector but a facility to inject data into a link that can flow to other regions.    Similar to what VectorFileSensor does with data from a file.  I am thinking this might be implemented as a kind of encoder.

> " and I think the main InputRegion "

Not following you on this statement.

breznak commented on py/htm/advanced/regions/RawValues.py:
```diff
@@ -0,0 +1,126 @@
+
+class RawValues(PyRegion):
+    """
+    RawDate is a simple region used to send raw scalar values into networks.
```
@dkeeney this sounds generally useful, do we want to promote the region from `.advanced` to `htm.regions` ? 

But that would be only for python. We already have `SensorRegion` which applies an encoder to the imputs. A solution would be to have an identity/pass-through encoder (=that does nothing with its inputs) and use that instead of this class. 
  - BaseEncoder could behave as an identity encoder.

dkeeney replied:
>  A solution would be to have an identity/pass-through encoder (=that does nothing with its inputs) and use that instead of this class.

Its the lack of a NetworkAPI region that is the problem.  This Raw region is one solution to it.  The EncoderRegion would have been able to do that if I could figure out how to make it work.

breznak replied:
Yes, I was thinking about a Generic encoder region #637 would come in handy here. 

So, the remaining point, do we want to move "RawValues" (maybe rename to ??) from advanced to the common py regions?

fcr replied:
Maybe RawSensor also. It queues sparse input and pops it as dense.
In either case, the `region.__init__.py` will need to move as well since it provides extractList.

fcr replied:
I am about to push the changes due to #670

fcr commented on py/htm/advanced/examples/thing_classification/l2l4l6_experiment.py:
```diff
@@ -0,0 +1,254 @@
+    Load Thing sensation data. There is one file per object, each row contains one
+    feature, location pairs. The format is as follows:
+        [(-33.6705, 75.5003, 2.4207)/10] => [[list of active bits of location],
+                                                                                 [list of active bits of feature]]
```
I think that Numenta massaged them to their own needs so an automatic download would not work.

fcr commented:
Thank you [@breznak](https://github.com/breznak) and [@dkeeney](https://github.com/dkeeney).
I am sure that is not easy to review such a large body of code.
If you want another part to port, htmresearch.core->ApicalTiebreakTemporalMemory.cpp would be a good candidate. 
This would be a great performance boost to the framework. After htm.core this should be child's play for you. ;-).

ctrl-z-9000-times commented:
Thank you @fcr, i ve tried several times to port this code and each time gotten lost in the woods. This is a great addition to this repo!
